lib/Sema/CSApply.cpp

```diff
@@ -1233,11 +1233,22 @@ namespace {
                          ConstraintLocatorBuilder locator,
                          bool isImplicit, AccessSemantics semantics) {
       // Determine the declaration selected for this subscript operation.
-      auto selected = getOverloadChoice(
+      auto selected = getOverloadChoiceIfAvailable(
                         cs.getConstraintLocator(
                           locator.withPathElement(
                             ConstraintLocator::SubscriptMember)));
-      auto choice = selected.choice;
+
+      // Handles situation where there was a solution available but it didn't
+      // have a proper overload selected from subscript call, might be because
+      // solver was allowed to return free or unresolved types, which can
+      // happen while running diagnostics on one of the expressions.
+      if (!selected.hasValue()) {
+        cs.TC.diagnose(base->getLoc(), diag::type_of_expression_is_ambiguous)
+             .highlight(base->getSourceRange());
+        return nullptr;
```

Comment:
@rudkx I've poked it a bit more from different angle, and it looks like even example with both parameters specified is going to be ambiguous, because original system would not produce any solutions for `TupleExpr` which contains parameters and failure diagnostics would start from re-checking "subscript" because it's a "function" in that case (when type-checking closure separately), it wouldn't be able to resolve $0 nor $1 (because there `TupleExpr` is not available in that context) and would fail with ambiguous result.

What I did to archive that is - i've added additional parameter to `typeCheckChildIndependently` which identifies if we are trying to re-check a closure and if so - don't allow unresolved types - such avoids failing with subscript too early and actually going and running failure diagnostics for sub-expressions of the closure containing subscript. Such allows to avoid diagnosing problem in `buildSubscript` directly, do you think it would be useful for you to take a look at that?